nixos/modules/services/networking/pppd.nix

```diff
@@ -0,0 +1,137 @@
+    # modern system, so this wrapper transparently moves those files
+    # into /run/pppd.
+    pppd-wrapped = pkgs.stdenv.mkDerivation {
+      name = "pppd-wrapped";
+      phases = [ "installPhase" ];
+      buildInputs = [ pkgs.makeWrapper ];
+      installPhase = ''
+        mkdir -p $out/sbin
+        makeWrapper ${getBin config.services.pppd.package}/sbin/pppd $out/sbin/pppd \
+          --set LD_PRELOAD "${pkgs.libredirect}/lib/libredirect.so" \
+          --set NIX_REDIRECTS "/var/run=/run/pppd"
+      '';
+    };
+
+    enabledConfigs = filter (f: f.enable) (attrValues config.services.pppd.peers);
+
+    mkEtc = cfg: {
+      "ppp/peers/${cfg.name}".text = cfg.config;
+    };
+
+    mkSystemd = cfg: {
+      "pppd-${cfg.name}" = {
+        restartTriggers = [ config.environment.etc."ppp/peers/${cfg.name}".source ];
+        before = [ "network.target" ];
+        wants = [ "network.target" ];
+        after = [ "network-pre.target" ];
+        serviceConfig = {
+          ExecStart = "${getBin pppd-wrapped}/sbin/pppd call ${cfg.name} nodetach nolog";
+          Restart = "always";
+          RestartSec = 5;
+
+          AmbientCapabilities = "CAP_SYS_TTY_CONFIG CAP_NET_ADMIN CAP_NET_RAW CAP_SYS_ADMIN";
+          CapabilityBoundingSet = "CAP_SYS_TTY_CONFIG CAP_NET_ADMIN CAP_NET_RAW CAP_SYS_ADMIN";
+          LockPersonality = true;
+          MemoryDenyWriteExecute = true;
+          NoNewPrivileges = true;
+          PrivateMounts = true;
+          PrivateTmp = true;
+          ProtectControlGroups = true;
+          ProtectHome = true;
+          ProtectHostname = true;
+          ProtectKernelModules = true;
+          # pppd can be configured to tweak kernel settings.
+          ProtectKernelTunables = false;
+          ProtectSystem = "strict";
+          RemoveIPC = true;
+          RestrictAddressFamilies = "AF_PACKET AF_UNIX AF_PPPOX AF_ATMPVC AF_ATMSVC AF_INET AF_INET6 AF_IPX";
+          RestrictNamespaces = true;
+          RestrictRealtime = true;
+          RestrictSUIDSGID = true;
+          SecureBits = "no-setuid-fixup-locked noroot-locked";
+          SystemCallArchitectures = "native";
+
+          # All pppd instances on a system must share a runtime
+          # directory in order for PPP multilink to work correctly. So
+          # we give all instances the same /run/pppd directory to store
+          # things in.
+          #
+          # For the same reason, we can't set PrivateUsers=true, because
+          # all instances need to run as the same user to access the
+          # multilink database.
+          RuntimeDirectory = "pppd";
+          RuntimeDirectoryPreserve = true;
+        };
+        wantedBy = optional cfg.autostart "multi-user.target";
+      };
+    };
+
+    etcFiles = map mkEtc enabledConfigs;
+    systemdConfigs = map mkSystemd enabledConfigs;
+
+  in {
+    environment.etc = mkMerge etcFiles;
+    systemd.services = mkMerge systemdConfigs;
+  };
```

Comment:
Should the package be added to the environment or is that not necessary (i.e if it is only run by systemd)

Reply:
I don't know what the custom is for NixOS. The only reason to invoke `pppd` directly is debugging a connection, but you can do that easily with `systemctl restart pppd-peer` as well. Based on that, I thought I would not pollute the system $PATH with a "useless" daemon, but it's not a strong opinion. WDYT?

Reply:
The only reason I would consider it is that adding `pppd` to `environment.systemPackages` won't add your wrapped version (which I'm guessing is needed to debug). It is probably fine not to have it

Reply:
Good point!

IMO, it's still easy to debug with `systemctl restart pppd-name`, so having `pppd` or the wrapper in `systemPackages` is just more noise for nothing. And it's easier to add it later if we're wrong, than to remove it and potentially break someone who depends on it. So I propose we start without adding to `systemPackages`, and see how that goes.